src/csrc/alphafold/fused_gating_attention_fwd_tmpl.h

```diff
@@ -152,154 +152,231 @@ float alpha = (1.0 / sqrt(key_dim));
+// 764, 764]  = [512, 764, 8, 32] * [512, 764, 8, 32] + [512, 1, 1, 764] */ if
+// (nonbatched_bias.size(0) > 0)
+//     logits = at::add(logits, at::unsqueeze(nonbatched_bias, 0)); /* [512, 8,
+//     764, 764]  = [512, 8, 764, 764] + [1, 8, 764, 764] */
+// weights = at::_softmax(logits, -1, false); /* [512, 8, 764, 764] = [512, 8,
+// 764, 764] */ auto weighted_avg = at::einsum("bhqk,bkhc->bqhc", {weights,
+// v}).contiguous();          /* [512, 764, 8, 32]  = [512, 8, 764, 764] * [512,
+// 764, 8, 32] */
 
 auto a_zero_tpp = SCOPEIT(SetZeroTPP<T>(A_BLOCKSIZE * H_t), EW_ZERO);
+auto a_cpy_tpp = SCOPEIT(CpyTPP<T>(A_BLOCKSIZE, H_t, N_t* H_t, H_t), EW_COPY);
 auto a_cpy2_tpp = SCOPEIT(CpyTPP<T>(A_BLOCKSIZE, H_t, H_t, N_t* H_t), EW_COPY);
-// auto a_cpy3_tpp = SCOPEIT(CpyTPP<T>(A_BLOCKSIZE, H_t), EW_COPY);
-// auto a_cpy4_tpp = SCOPEIT(CpyTPP<T>(A_BLOCKSIZE, Ak_BLOCKSIZE, S_t,
-// Ak_BLOCKSIZE), EW_COPY);
 
-auto a_brgemm_tpp = SCOPEITGEMM(
+if (S_t < 2048){
+
+  auto a_brgemm_tpp = SCOPEITGEMM(
     (BrgemmTPP<
         T,
         T>(A_BLOCKSIZE, A_BLOCKSIZE, H_t, 0, 0, H_t, S_t, S_t, 0.0, 0, 1)));
 
-auto a_brgemm2_tpp = SCOPEITGEMM((BrgemmTPP<T, T>(
+  auto c_brgemm_tpp = SCOPEITGEMM((BrgemmTPP<T, T>(
+      A_BLOCKSIZE,
+      H_t,
+      A_BLOCKSIZE,
+      A_BLOCKSIZE,
+      A_BLOCKSIZE* N_t* H_t,
+      S_t,
+      N_t* H_t,
+      H_t,
+      1.0,
+      0,
+      1)));
+
+  auto a_addbias_tpp = SCOPEIT(AddBiasTPP<T>(A_BLOCKSIZE, S_t, S_t), BIAS);
+  auto a_add_nbbias_tpp =
+      SCOPEIT((AddTPP<T, T>(A_BLOCKSIZE, S_t, S_t, S_t)), BIAS);
+
+  auto a_add_sfmask_tpp =
+      SCOPEIT(AddBiasTPP<T>(A_BLOCKSIZE, S_t - Sp_t, S_t), BIAS);
+  auto a_softmax_tpp =
+      SCOPEIT((VarSoftMaxFwdTPP<float, T>(A_BLOCKSIZE, S_t)), SOFTMAX);
+
+  {
+    RECORD_SCOPE(alpha_ac_gemm, {q, k, bias});
+    {
+      RECORD_FUNCTION("parallel_for", std::vector<c10::IValue>());
+  #pragma omp parallel for collapse(3)
+      for (int i = 0; i < B_t; i++) {
+        for (int n = 0; n < N_t; n++) {
+          for (int j1 = 0; j1 < S_t; j1 += A_BLOCKSIZE) {
+            T tmp_qv[A_BLOCKSIZE * H_t];
+            T tmp_logits[A_BLOCKSIZE][S_t];
+
+            a_cpy_tpp(&q_a[i][j1][n][0], &tmp_qv[0]);
+            for (int j2 = 0; j2 < S_t; j2 += A_BLOCKSIZE) {
+              a_brgemm_tpp(
+                  &tmp_qv[0], &k_a[i][n * H_t * S_t + j2], &tmp_logits[0][j2], 1);
+            }
+
+            a_addbias_tpp(&bias_a[i][0], &tmp_logits[0][0]);
+            if (flag)
+              a_add_nbbias_tpp(
+                  &nonbatched_bias_a[0][n][j1][0],
+                  &tmp_logits[0][0],
+                  &tmp_logits[0][0]);
+
+            if (S_t == Sp_t) {
+              a_softmax_tpp(1, &tmp_logits[0][0], &tmp_logits[0][0]);
+            } else {
+              a_add_sfmask_tpp(&sfmask_a[0][0], &tmp_logits[0][Sp_t]);
+              a_softmax_tpp(1, &tmp_logits[0][0], &tmp_logits[0][0]);
+            }
+
+            a_zero_tpp(&tmp_qv[0]);
```

Comment:
Why use beta=1.0 and separate zero_tpp? Why not simply set beta=0.0 in c_brgemm_tpp?

Reply:
removed zero tpp and used beta=0.0 in the latest commit